lms/djangoapps/bookmarks_api/views.py

```diff
@@ -0,0 +1,210 @@
+            return Response(
+                {
+                    "developer_message": error_message,
+                    "user_message": error_message
+                },
+                status=status.HTTP_400_BAD_REQUEST
+            )
+
+        usage_id = request.DATA.get('usage_id', None)
+        if not usage_id:
+            error_message = _('No usage id provided for bookmark')
+            return Response(
+                {
+                    "developer_message": error_message,
+                    "user_message": error_message
+                },
+                status=status.HTTP_400_BAD_REQUEST
+            )
+
+        try:
+            usage_key = UsageKey.from_string(usage_id)
+
+            # usage_key's course_key may have an empty run property
+            usage_key = usage_key.replace(course_key=modulestore().fill_in_run(usage_key.course_key))
+            course_key = usage_key.course_key
+        except InvalidKeyError:
+            error_message = _(u"invalid usage id '{usage_id}'".format(usage_id=usage_id))
+            return Response(
+                {
+                    "developer_message": error_message,
+                    "user_message": error_message
+                },
+                status=status.HTTP_400_BAD_REQUEST
+            )
+
+        try:
+            module_item = modulestore().get_item(usage_key)
+
+            parent = module_item.get_parent()
+            location_path = []
+            while parent is not None:
+                display_name = parent.display_name if parent.display_name else _("(Unnamed)")
+                location_path.append({"display_name": display_name, "usage_id": unicode(parent.location)})
+                parent = parent.get_parent()
+        except ItemNotFoundError:
+            log.warn(
+                "Invalid location for usage_id: {usage_id}".format(
+                    usage_id=usage_id
+                )
+            )
+            raise Http404
+
+        path_list = location_path[:2] if location_path else list()
+        path_list.reverse()
+        bookmarks_dict = {
+            "usage_key": usage_key,
+            "course_key": course_key,
+            "user": request.user,
+            "display_name": module_item.display_name,
+            "_path": path_list
+        }
+        try:
+            bookmark = Bookmark.create(bookmarks_dict)
+        except ValidationError as error:
+            log.debug(error, exc_info=True)
+            return Response(status=status.HTTP_400_BAD_REQUEST)
+
+        return Response(serializers.BookmarkSerializer(bookmark).data, status=status.HTTP_201_CREATED)
+
+
+class BookmarksDetailView(APIView):
+    """
+    List all bookmarks or create.
+    """
+    authentication_classes = (OAuth2Authentication, SessionAuthentication)
+    permission_classes = (permissions.IsAuthenticated,)
+
+    serializer_class = serializers.BookmarkSerializer
+
+    def get(self, request, username=None, usage_key_string=None):
```

Comment:
I prefer `usage_id` for the string representation and `usage_key` for the object.